tests/prototypes/prototype_mini.py

```diff
@@ -0,0 +1,472 @@
+            if self.lapse
+            else 0
+        )
+
+    @property
+    def start_timestamp(self) -> float:
+        if self._start_timestamp is None:
+            raise TypeError(
+                "Start-timestamp is not available. ``start`` was never called."
+            )
+        else:
+            return self._start_timestamp
+
+    @property
+    def stop_timestamp(self) -> float:
+        if self._stop_timestamp is None:
+            raise TypeError(
+                "Stop-timestamp is not available. ``stop`` was never called."
+            )
+        else:
+            return self._stop_timestamp
+
+    def start(self) -> None:
+        import time
+
+        try:
+            self.start_timestamp
+        except TypeError:
+            self._start_timestamp = time.time()
+        else:
+            raise RuntimeError(
+                "Start-timestamp is already recorded. "
+                "``start`` cannot be called twice."
+            )
+
+    def stop(self) -> None:
+        import time
+
+        try:
+            self.start_timestamp
+        except TypeError:
+            raise RuntimeError("``start`` should be called before ``stop``.")
+
+        try:
+            self.stop_timestamp
+        except TypeError:
+            self._stop_timestamp = time.time()
+        else:
+            raise RuntimeError(
+                "Stop-timestamp is already recorded. "
+                "``stop`` cannot be called twice."
+            )
+
+    def lap(self, app_name: str) -> None:
+        import time
+
+        app_lapse = self.lapse.setdefault(app_name, [])
+        app_lapse.append(time.time())
+
+
+class BaseApp(LogMixin, ABC):
+    logger: BeamlimeLogger
+    stop_watch: StopWatch
+    target_counts: TargetCounts
+
+    @property
+    def app_name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def target_count_reached(self) -> bool:
+        return self.target_counts <= self.data_counts
+
+    def __init__(self) -> None:
+        self.data_counts: int = 0
+        super().__init__()
+
+    async def commit_process(self):
+        self.stop_watch.lap(self.app_name)
+        self.data_counts += 1
+        await asyncio.sleep(0)
+
+    def data_pipe_monitor(
+        self,
+        pipe: List[Any],
+        timeout: float = 5,
+        interval: float = 1 / 14,
+        prefered_size: int = 1,
+        target_size: int = 1,
+    ):
+        from beamlime.core.schedulers import async_retry
+
+        @async_retry(
+            TimeoutError, max_trials=int(timeout / interval), interval=interval
+        )
+        async def wait_for_preferred_size() -> None:
+            if len(pipe) < prefered_size:
+                raise TimeoutError
+
+        async def is_pipe_filled() -> bool:
+            try:
+                await wait_for_preferred_size()
+            except TimeoutError:
+                ...
+            return len(pipe) >= target_size
+
+        return is_pipe_filled
```

Comment:
It should return the function object `is_pipe_filled`.

It has triple steps, 
in `data_pipe_monitor`
1. `wait_for_preferred_size` will be re-run until it reaches the condition, `len(pipe) < preferred_size` in this case, and throw an error if it does't meet the condition.
2. `is_pipe_filled` will use `wait_for_preferred_size` to wait until the `pipe` is filled up to `preferred size`, and returns True/False based on the `target_size`.

`target_size` is what you `need` at the minimum and `preferred_size` is what you want to wait for.
For example, you might want to wait until there are 10 pieces of data in the list but you only need 1 to proceed.
At the end `data_pipe_monitor` will give you this handle `is_pipe_filled` to see if the `pipe` object is filled with `target_size` pieces of data.

3. `data_pipe_monitor` create two functions above and return an async function `is_pipe_filled` so that it could be awaited in while loop or in an if condition.

`data_pipe_monitor` could also just handle the error and return `len(pipe) >= target_size` but then
it has to create `wait_for_preferred_size` every time it checks the size of the pipe so I wrapped it this way.

Does this make sense....?